doc/source/rllib/env-to-module-connector.rst

```diff
@@ -0,0 +1,546 @@
+
+    config = (
+        PPOConfig()
+
+        # Configure a simple 2x2 grid-world.
+        # ____
+        # |S |  <- S=start position
+        # | G|  <- G=goal position
+        # ----
+        .environment("FrozenLake-v1", env_config={"desc": ["SF", "FG"]})
+
+        # Plug your custom connector piece into the env-to-module pipeline.
+        .env_runners(
+            env_to_module_connector=(
+                lambda env, spaces, device: IntObservationToOneHotTensor()
+            ),
+        )
+    )
+    algo = config.build()
+    # Train one iteration.
+    print(algo.train())
+
+
+.. _observation-preprocessors-adding-rewards-to-obs:
+
+Adding recent rewards to the batch
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Assume you wrote a custom :ref:`RLModule <rlmodule-guide>` that requires the last three received
+rewards as input in the calls to any of its `forward_..()` methods.
+
+You can use the same :py:class:`~ray.rllib.connectors.env_to_module.observation_preprocessor.SingleAgentObservationPreprocessor`
+API to achieve this.
+
+In the following example, you extract the last three rewards from the ongoing episode and concatenate
+them with the observation to form a new observation tensor.
+Note that you also have to change the observation space returned by the connector, since
+there are now three more values in each observation:
+
+
+.. testcode::
+
+    import gymnasium as gym
+    import numpy as np
+
+    from ray.rllib.connectors.env_to_module.observation_preprocessor import SingleAgentObservationPreprocessor
+
+
+    class AddPast3Rewards(SingleAgentObservationPreprocessor):
+        """Extracts last 3 rewards from episode and concatenates them to the observation tensor."""
+
+        def recompute_output_observation_space(self, in_obs_space, in_act_space):
+            # Based on the input observation space (), return the output observation
+            # space. Implementing this method is crucial for the pipeline to know its output
+            # spaces, which are an important piece of information to construct the succeeding
+            # RLModule.
+
+            assert isinstance(in_obs_space, gym.spaces.Box) and len(in_obs_space.shape) == 1
+            return gym.spaces.Box(-100.0, 100.0, (in_obs_space.shape[0] + 3,), np.float32)
+
+        def preprocess(self, observation, episode):
+            # Extract the last 3 rewards from the ongoing episode using a python `slice` object.
+            # Alternatively, you can also pass in a list of indices, [-3, -2, -1].
+            past_3_rewards = episode.get_rewards(indices=slice(-3, None))
+
+            # Concatenate the rewards to the actual observation.
+            new_observation = np.concatenate([
+                observation, np.array(past_3_rewards, np.float32)
+            ])
+
+            # Return the new observation.
+            return new_observation
+
+
+Note that the preceding example should work without any further action required on your model,
```

Comment:
This note here is so important that it deserves an extra box. A user should know, that the `RLModule`'s default encoder (different in custom ones) is built upon the connector pipeline's output observation space.

Reply:
done